File: server/types/Point.py
# -*- coding: utf-8 -*-

import math

HEADING_TABLE_X = (0, 1, 1, 1, 0, -1, -1, -1)
HEADING_TABLE_Y = (-1, -1, 0, 1, 1, 1, 0, -1)

class Point():

    def __init__(self, x = None, y = None):
        self._x = 0
        self._y = 0

        if x:
            self._x = x

        if y:
            self._y = y

    def set(self, x, y):
        self._x = x
        self._y = y

    def forward(self, heading):
        self._x += HEADING_TABLE_X[heading]
        self._y += HEADING_TABLE_Y[heading]

    def backward(self, heading):
        self._x -= HEADING_TABLE_X[heading]
        self._y -= HEADING_TABLE_Y[heading]

    def getLineDistance(self, pt):
        diffX = pt._x - self._x
        diffY = pt._y - self._y
        return math.sqrt(diffX * diffX + diffY * diffY)

    def getTileLineDistance(self, pt):
        return max(math.fabs(pt._x - self._x), math.fabs(pt._y - self._y))

    def getTileDistance(self, pt):
        return math.fabs(pt._x - self._x) + math.fabs(pt._y - self._y)

    def isInScreen(self, pt):
        dist = self.getTileDistance(pt)

        if dist > 17:
            return False
        elif dist <= 13:
            return True
        else:
            dist = math.fabs(pt._x - (self._x - 15)) + math.fabs(pt._y - (self._y - 15))
            if 17 <= dist and dist <= 43:
                return True
            return False

    def __eq__(self, other):
        return other._x == self._x and other._y == self._y

    def hashCode(self):
        return 7 * self._x + self._y

File: Module/Point.cpp
//
// Created by 范炜东 on 2018/12/13.
//

#include <cmath>
#include <algorithm>
#include <Poco/Format.h>
#include "Point.h"

namespace Lineage {

Point::Point() : x(0), y(0) {
}

Point::Point(int xPos, int yPos) : x(xPos), y(yPos) {
}

Point::~Point() {
}

void Point::backward(Towards t) {
    x -= TOWARDS_TABLE_X[t];
    y -= TOWARDS_TABLE_Y[t];
}

void Point::forward(Towards t) {
    x += TOWARDS_TABLE_X[t];
    y += TOWARDS_TABLE_Y[t];
}

double Point::getLineDistance(const Point &pt) {
    long diffX = pt.x - x;
    long diffY = pt.y - y;
    return std::sqrt(diffX * diffX + diffY * diffY);
}

int32_t Point::getTileDistance(const Point &pt) {
    int diffX = pt.x - x;
    int diffY = pt.y - y;
    return std::max(std::abs(diffX), std::abs(diffY));
}

int32_t Point::getTileLineDistance(const Point &pt) {
    int diffX = pt.x - x;
    int diffY = pt.y - y;
    return std::abs(diffX) + std::abs(diffY);
}

int32_t Point::hashCode() {
    return 7 * x + y;
}

bool Point::isInScreen(const Point &pt) {
    int dist = getTileDistance(pt);
    // 首先使用平行距离进行判断，过滤掉那些太远影响近似计算的点
    if (dist > 17) {
        return false;
    } else if (dist <= 13) // 简化判断
    {
        return true;
    } else {
        // 判断两点间的直线距离是否在范围内的近似算法，首先以当前坐标为原点建立坐标系，
        // 然后计算沿着X轴和Y轴移动到目标的距离
        int diffX = pt.x - (x - 15);
        int diffY = pt.y - (y - 15);
        int dist = std::abs(diffX) + std::abs(diffY);
        if (17 <= dist && dist <= 43)
            return true;
    }

    return false;
}

bool Point::operator==(const Point &pt) {
    return (pt.x == x && pt.y == y);
}

bool Point::operator!=(const Point &pt) {
    return (pt.x != x || pt.y != y);
}

std::string Point::toString() {
    return Poco::format("(%d, %d)", x, y);
}

}